### ISSUE TITLE
ExpressionGenerator: using FactManager to help finding numbers for the addition value

### ISSUE THREAD

**Issue by jiradeto** [google/graphicsfuzz#663]
When finding two numbers whose sum is equal to Value, we can involve FactManager to bias choosing at least one number for which we have a known fact.

E.g., if we have facts about variables or functions that compute 12, 16, 3, 6, 2, and we're trying to generate 100, then it would be better if we can choose 12 + 88, or 94 + 6, or whatever.

**Comment by afd** [comment]
Perhaps we should equip the fact manager with a method that, given a type, returns the list of all distinct values of that type for which it knows at least one fact.

Then with e.g. 50% probability we could choose one of the summands from this list of values (if the list is non-empty).

**Comment by jiradeto** [comment]
I don't quite follow:

> Perhaps we should equip the fact manager with a method that, given a type, returns the list of all distinct values of that type for which it knows at least one fact.
> 

Why do we need to consider knowing _at least one fact_? It seems to me that even if we have all unknown values (Optional.empty()) in the return list, we still can find two numbers anyway?

For example, if we are going to generate 100, and all distinct values in the list return from the Fact manager are all unknown of Integer type, we thus call the Generate Expression to, given the correct type, generates any arbitrary value. Let's say the generator eventually generates 56 for this unknown value, so we subtract the original value by this number and finally derive the summands: 100 = 56 + 44;

Please correct me if I am wrong.

**Comment by afd** [comment]
I agree that if we require generating some integer value x, we can always pick two integers y and z such that y + z = x.

However, suppose what we are doing is generating the body of:

int must_return_42(int known_to_be_22, int known_to_be_5) {
   ...
}

and suppose we choose to implement this as:

return _ + _;

Our task is therefore to choose two numbers for the "_" slots.  If we choose these at random, it is extremely unlikely that we will pick 22 or 5.  And if we do *not* pick 22 or 5, we can't make use of the parameters "known_to_be_22" or "known_to_be_5".  Also, if we previously generated functions or global variables that are known to compute or hold particular values, we are very unlikely to randomly choose values for "_" above that allow us to make use of these.

My proposal is that when we select the values for the sum, we bias our choice towards selecting - for one or other of the operands - a value for which we know a fact.  Say we choose 22 for the first "_" above.  This means we can do:

return known_to_be_22 + (... guaranteed to evaluate to 20 ...)

and if we end up expanding (... guaranteed to evaluate to 20 ...) to another "_ + _" then we might choose known_to_be_7 for the second "_", so that we have:

return known_to_be_22 + ((... guaranteed to evaluate to 13 ...) + known_to_be_7)

which might finally expand to something like:

return known_to_be_22 + ((known_to_be_7 + 6) + known_to_be_7)

This is a nice expression because it relies on the guarantees provided by the arguments to the function.  But we would be extremely unlikely to generated it purely randomly if we did not bias the choice of values towards values about which the fact manager knows something.

**Comment by jiradeto** [comment]
Thanks for your detailed explanation and example. 

When I was reading _"returns the list of all distinct values of that type for which it knows at least one fact"_  there were 2 ideas came into my mind: 

1. Returns a list of values in which every element is known by Fact Manager (All values are not Optional.empty()).
For example, the items in this list are 
```
[NumericValue(BasicType.INT, Optional.of(2),  
NumericValue(BasicType.INT, Optional.of(32), 
NumericValue(BasicType.INT, Optional.of(19)]
```

2. Returns a list of all values holds by Fact manager with the correct type. **And this list must contain at least one known value**. 
For example, the items in this list are 
```
[NumericValue(BasicType.INT, Optional.empty(),
 NumericValue(BasicType.INT, Optional.of(2), // Knowing just only one value is OK
 NumericValue(BasicType.INT, Optional.empty()].
```


After reading your comment I believe you mean no.1?

**Comment by afd** [comment]
For the use case of choosing operands for _ + _, you are correct that (1) is what we require.

However, it might be simplest, from an API perspective, to give the fact manager a method that can return a list of all the values -- whether known or not -- for which it holds facts.  Then we can filter this list based on more specific requirements, such as discarding all of the values that are unknown.

The reason is that "known" is a bit subtle when we consider composite types.  For example, the fact manager might have:

CompositeValue(VEC4, UNKNOWN) // completely unknown vec4
CompositeValue(VEC4, [ (FLOAT, 2.0), (FLOAT, 4.0), (FLOAT, 2.1), (FLOAT, 1.5) ]) // completely known vec4
CompositeValue(VEC4, [ (FLOAT, 2.0), UNKNOWN, (FLOAT, 2.1), (FLOAT, 1.5) ]) // partially known vec4

So having a way to say: "Give me all the distinct values for which you know facts" seems to be the most general thing we can add to the fact manager.

**Comment by jiradeto** [comment]
Fixed by #681
